### PR TITLE
Heat maps [WIP]

### LINKED ISSUE #3206: Request for Google Maps API heat maps
It would be really useful to be able to use the heatmap feature of Google Maps to visualise data - ideally the ability to represent a "weighting" at a particular lat/long in much the same way as the region maps currently have a "metric" associated with a region:

https://developers.google.com/maps/documentation/javascript/heatmaplayer

Thanks for all your hard work on developing Metabase this far - it's a very useful tool! :+1: :smiley:

### PR DESCRIPTION
**Pull request by tlrobinson**
TODO:

- [x] Settings UX (currently adds a "pin map type" setting, but should probably be a top level map type?)
- [ ] Tweak various heat map parameters (based on zoom level?)
- [ ] backend should return the bin size in the column object (similar to date/time's `unit`)
- [ ] it would be nice if our sample dataset had more realistic data, especially for demonstrating drill down etc.
- [ ] better colors
- [ ] Check that this isn't affecting us now that we're using leaflet.draw as well https://github.com/Leaflet/Leaflet.heat/issues/80

Resolves #3206

Adds two styles of heat maps (not sure what the correct names are)

screenshot 2017-04-04 23 23 07
screenshot 2017-04-04 23 22 59

Builds on @senior's histogram PR #4658 for the backend. Hacked it up a bit to allow for both lat/lon breakouts (was crashing previously), but still needs a lot of work (e.x. if breakout is also filtered use that as min/max, also, don't bin the field in the filters, maybe allow configuration of # of bins or bin size in MBQL/frontend?)

### CODE REVIEW

**Comment by mazameli** [comment]
Hmm I'm getting `java.lang.NullPointerException` when trying to run the same query you have in your screenshots.

**Comment by tlrobinson** [comment]
@maz try syncing the database after checking out this branch and restarting the backend

File: resources/migrations/052_add_min_max_to_fields.yaml [hunk 1]
```diff
@@ -0,0 +1,21 @@
+databaseChangeLog:
+  - changeSet:
+      id: 52
```
Comment (camsaul) [inline resources/migrations/052_add_min_max_to_fields.yaml]: we'll want to change this to 55 before merging

**Comment by mazameli** [comment]
Still can't seem to get this branch to work sadly. I'm trying to just see the raw data of the People table in the sample dataset, and still seeing `java.lang.NullPointerException`, along with a number of JS errors:

```
warning.js?5047:40 Warning: Failed prop type: Invalid prop `error` of type `string` supplied to `VisualizationError`, expected `object`.
    in VisualizationError (created by QueryVisualization)
    in QueryVisualization (created by LegacyQueryBuilder)
    in div (created by LegacyQueryBuilder)
    in div (created by LegacyQueryBuilder)
    in div (created by LegacyQueryBuilder)
    in LegacyQueryBuilder (created by QueryBuilder)
    in div (created by QueryBuilder)
    in QueryBuilder (created by QueryBuilder)
    in QueryBuilder (created by RouterContext)
    in Unknown (created by UserIsAuthenticated(Component))
    in UserIsAuthenticated(Component) (created by Connect(UserIsAuthenticated(Component)))
    in Connect(UserIsAuthenticated(Component)) (created by MetabaseIsSetup(Connect(UserIsAuthenticated(Component))))
    in MetabaseIsSetup(Connect(UserIsAuthenticated(Component))) (created by Connect(MetabaseIsSetup(Connect(UserIsAuthenticated(Component)))))
    in Connect(MetabaseIsSetup(Connect(UserIsAuthenticated(Component)))) (created by RouterContext)
    in div (created by App)
    in App (created by App)
    in App (created by RouterContext)
    in RouterContext (created by Router)
    in Router
    in Provider
printWarning @ warning.js?5047:40
warning @ warning.js?5047:64
checkReactTypeSpec @ checkReactTypeSpec.js?3c1d:84
validatePropTypes @ ReactElementValidator.js?e05f:155
createElement @ ReactElementValidator.js?e05f:198
render @ QueryVisualization.jsx:188
(anonymous) @ ReactCompositeComponent.js?0b22:800
measureLifeCyclePerf @ ReactCompositeComponent.js?0b22:79
_renderValidatedComponentWithoutOwnerOrContext @ ReactCompositeComponent.js?0b22:799
_renderValidatedComponent @ ReactCompositeComponent.js?0b22:826
_updateRenderedComponent @ ReactCompositeComponent.js?0b22:750
_performComponentUpdate @ ReactCompositeComponent.js?0b22:728
updateComponent @ ReactCompositeComponent.js?0b22:649
receiveComponent @ ReactCompositeComponent.js?0b22:551
receiveComponent @ ReactReconciler.js?54ca:129
updateChildren @ ReactChildReconciler.js?0bf2:113
_reconcilerUpdateChildren @ ReactMultiChild.js?dd5a:212
_updateChildren @ ReactMultiChild.js?dd5a:316
updateChildren @ ReactMultiChild.js?dd5a:303
_updateDOMChildren @ ReactDOMComponent.js?2aca:940
updateComponent @ ReactDOMComponent.js?2aca:758
receiveComponent @ ReactDOMComponent.js?2aca:720
receiveComponent @ ReactReconciler.js?54ca:129
updateChildren @ ReactChildReconciler.js?0bf2:113
_reconcilerUpdateChildren @ ReactMultiChild.js?dd5a:212
_updateChildren @ ReactMultiChild.js?dd5a:316
updateChildren @ ReactMultiChild.js?dd5a:303
_updateDOMChildren @ ReactDOMComponent.js?2aca:940
updateComponent @ ReactDOMComponent.js?2aca:758
receiveComponent @ ReactDOMComponent.js?2aca:720
receiveComponent @ ReactReconciler.js?54ca:129
updateChildren @ ReactChildReconciler.js?0bf2:113
_reconcilerUpdateChildren @ ReactMultiChild.js?dd5a:212
_updateChildren @ ReactMultiChild.js?dd5a:316
updateChildren @ ReactMultiChild.js?dd5a:303
_updateDOMChildren @ ReactDOMComponent.js?2aca:940
updateComponent @ ReactDOMComponent.js?2aca:758
receiveComponent @ ReactDOMComponent.js?2aca:720
receiveComponent @ ReactReconciler.js?54ca:129
_updateRenderedComponent @ ReactCompositeComponent.js?0b22:758
_performComponentUpdate @ ReactCompositeComponent.js?0b22:728
updateComponent @ ReactCompositeComponent.js?0b22:649
receiveComponent @ ReactCompositeComponent.js?0b22:551
receiveComponent @ ReactReconciler.js?54ca:129
updateChildren @ ReactChildReconciler.js?0bf2:113
_reconcilerUpdateChildren @ ReactMultiChild.js?dd5a:212
_updateChildren @ ReactMultiChild.js?dd5a:316
updateChildren @ ReactMultiChild.js?dd5a:303
_updateDOMChildren @ ReactDOMComponent.js?2aca:940
updateComponent @ ReactDOMComponent.js?2aca:758
receiveComponent @ ReactDOMComponent.js?2aca:720
receiveComponent @ ReactReconciler.js?54ca:129
_updateRenderedComponent @ ReactCompositeComponent.js?0b22:758
_performComponentUpdate @ ReactCompositeComponent.js?0b22:728
updateComponent @ ReactCompositeComponent.js?0b22:649
receiveComponent @ ReactCompositeComponent.js?0b22:551
receiveComponent @ ReactReconciler.js?54ca:129
_updateRenderedComponent @ ReactCompositeComponent.js?0b22:758
_performComponentUpdate @ ReactCompositeComponent.js?0b22:728
updateComponent @ ReactCompositeComponent.js?0b22:649
receiveComponent @ ReactCompositeComponent.js?0b22:551
receiveComponent @ ReactReconciler.js?54ca:129
_updateRenderedComponent @ ReactCompositeComponent.js?0b22:758
_performComponentUpdate @ ReactCompositeComponent.js?0b22:728
updateComponent @ ReactCompositeComponent.js?0b22:649
receiveComponent @ ReactCompositeComponent.js?0b22:551
receiveComponent @ ReactReconciler.js?54ca:129
_updateRenderedComponent @ ReactCompositeComponent.js?0b22:758
_performComponentUpdate @ ReactCompositeComponent.js?0b22:728
updateComponent @ ReactCompositeComponent.js?0b22:649
receiveComponent @ ReactCompositeComponent.js?0b22:551
receiveComponent @ ReactReconciler.js?54ca:129
_updateRenderedComponent @ ReactCompositeComponent.js?0b22:758
_performComponentUpdate @ ReactCompositeComponent.js?0b22:728
updateComponent @ ReactCompositeComponent.js?0b22:649
receiveComponent @ ReactCompositeComponent.js?0b22:551
receiveComponent @ ReactReconciler.js?54ca:129
_updateRenderedComponent @ ReactCompositeComponent.js?0b22:758
_performComponentUpdate @ ReactCompositeComponent.js?0b22:728
updateComponent @ ReactCompositeComponent.js?0b22:649
receiveComponent @ ReactCompositeComponent.js?0b22:551
receiveComponent @ ReactReconciler.js?54ca:129
_updateRenderedComponent @ ReactCompositeComponent.js?0b22:758
_performComponentUpdate @ ReactCompositeComponent.js?0b22:728
updateComponent @ ReactCompositeComponent.js?0b22:649
performUpdateIfNecessary @ ReactCompositeComponent.js?0b22:565
performUpdateIfNecessary @ ReactReconciler.js?54ca:161
runBatchedUpdates @ ReactUpdates.js?3896:154
perform @ Transaction.js?0b4c:144
perform @ Transaction.js?0b4c:144
perform @ ReactUpdates.js?3896:93
flushBatchedUpdates @ ReactUpdates.js?3896:176
closeAll @ Transaction.js?0b4c:210
perform @ Transaction.js?0b4c:157
batchedUpdates @ ReactDefaultBatchingStrategy.js?5c73:66
enqueueUpdate @ ReactUpdates.js?3896:204
enqueueUpdate @ ReactUpdateQueue.js?dfdb:28
enqueueSetState @ ReactUpdateQueue.js?dfdb:213
ReactComponent.setState @ ReactComponent.js?345d:67
onStateChange @ connectAdvanced.js?e4e1:221
notify @ Subscription.js?c91a:34
notifyNestedSubs @ Subscription.js?c91a:69
onStateChange @ connectAdvanced.js?e4e1:214
dispatch @ createStore.js?e658:190
(anonymous) @ middleware.js?cf16:26
(anonymous) @ index.js?a032:93
(anonymous) @ index.js?1a83:32
(anonymous) @ index.js?e4e6:18
dispatch @ applyMiddleware.js?92e6:49
_callee$ @ redux.js?f3e4:16
tryCatch @ runtime.js?9bf3:68
invoke @ runtime.js?9bf3:303
prototype.(anonymous function) @ runtime.js?9bf3:120
step @ redux.js?f3e4:13
(anonymous) @ redux.js?f3e4:13
warning.js?5047:40 Warning: Each child in an array or iterator should have a unique "key" prop. Check the render method of `ShrinkableList`. See https://fb.me/react-warning-keys for more information.
    in div (created by ShrinkableList)
    in ShrinkableList (created by ShrinkableList)
    in ShrinkableList (created by QueryVisualization)
    in div (created by QueryVisualization)
    in div (created by QueryVisualization)
    in div (created by QueryVisualization)
    in QueryVisualization (created by LegacyQueryBuilder)
    in div (created by LegacyQueryBuilder)
    in div (created by LegacyQueryBuilder)
    in div (created by LegacyQueryBuilder)
    in LegacyQueryBuilder (created by QueryBuilder)
    in div (created by QueryBuilder)
    in QueryBuilder (created by QueryBuilder)
    in QueryBuilder (created by RouterContext)
    in Unknown (created by UserIsAuthenticated(Component))
    in UserIsAuthenticated(Component) (created by Connect(UserIsAuthenticated(Component)))
    in Connect(UserIsAuthenticated(Component)) (created by MetabaseIsSetup(Connect(UserIsAuthenticated(Component))))
    in MetabaseIsSetup(Connect(UserIsAuthenticated(Component))) (created by Connect(MetabaseIsSetup(Connect(UserIsAuthenticated(Component)))))
    in Connect(MetabaseIsSetup(Connect(UserIsAuthenticated(Component)))) (created by RouterContext)
    in div (created by App)
    in App (created by App)
    in App (created by RouterContext)
    in RouterContext (created by Router)
    in Router
    in Provider
```

**Comment by mazameli** [comment]
I'd love to check this out and help answer lingering UX questions to move this towards mergeability, but I can't get around the `java.lang.NullPointerException` error I'm still seeing when trying to load up raw data for the `People` table in the sample dataset. Syncing the db and restarting the backend isn't working.

**Comment by tlrobinson** [comment]
@mazameli try checking out this branch then reverting the backend to master: `git checkout master -- src`. The lat/lon binning won't work, so the grid style heatmap won't be useful, but you can check out the blobby (???) heatmap.

I think we should probably split this PR apart into two: 1) the blobby heatmap and 2) the grid heat map + backend support for binned lat/lon, since blobby heatmap works with non-binned data.

I think the main thing that needs to be done is to make the heatmap a top level map type setting, and tweak the heat map parameters (https://github.com/Leaflet/Leaflet.heat#lheatlayerlatlngs-options). I'm not sure yet if we'll be able to pick a static set of parameters or if we'll need to change them based on the zoom level and/or density of points.

**Comment by mazameli** [comment]
Hmm, when I do that I get `Exception in thread "main" java.io.FileNotFoundException: Could not locate taoensso/nippy__init.class or taoensso/nippy.clj on classpath., compiling:(metabase/models/interface.clj:1:1)` in the terminal when I do `lein run`.

**Comment by tlrobinson** [comment]
@mazameli I've removed the backend changes so this should work now. Do a clean checkout (e.x. `git checkout heat-maps; git fetch; git reset --hard origin/heat-maps`)

A few things to note:

1. I've moved "heat map" to a top level map type
2. I've temporarily added some of the heat map settings ([radius, blur, minOpacity, maxZoom](https://github.com/Leaflet/Leaflet.heat#lheatlayerlatlngs-options)) that will probably need to be tweaked, and maybe exposed as settings. However, I think these settings may need to be computed dynamically based on the "density" of data:
     * number of points in results
     * number of points visible in viewport? not sure how it will feel if that changes as you zoom in
3. I've hidden the "grid" heat map for now since we don't have binning yet (#4658).'
    * We could fall back to doing the binning client-side like we do for unaggregated graphs (and once it's implemented on the backend then show the little warning triangle like we do on graphs for unaggregated data)
    * bonus points for hexagonal binning https://www.mapbox.com/blog/binning-alternative-point-maps/
4. I added an optional metric field for heat maps, which we could also add to pin maps as the pin size (I imagine we'd want to use circles instead of actual pin icons)

**Comment by mazameli** [comment]
Yay, got this to work! (I'd forgotten to check it again.) The blob map seems cool.

> 1. I've moved "heat map" option to a top level map type

Yep, makes sense.

> 2. I've temporarily added some of the heat map settings ([radius, blur, minOpacity, maxZoom](https://github.com/Leaflet/Leaflet.heat#lheatlayerlatlngs-options)) that will probably need to be tweaked, and maybe exposed as settings. However, I think these settings may need to be computed dynamically based on the "density" of points:

That sounds good to me. I definitely was confused by the parameters. Obviously if we compute them dynamically based on the density of points, the settings don't make sense to expose in their current state.

> 4. I added an optional metric field for heat maps, which we could also add to pin maps as the pin size (I imagine we'd want to use circles instead of actual pin icons)

What does this do for heat maps? Does it also change the size of the blobs?

Also: does this require latitude and longitude fields? I guess this is a question for pin maps as well, but in the future I imagine we'll want to support 1D results, like count of rows by coordinate (lat+long fused into a single field), or count of rows by zip code, etc.

**Comment by 5an1ty** [comment]
Looks, great! Will see if I can get this to work on local machine.

**Comment by camsaul** [comment]
Closing this and other old POC or WIP PRs for now to help clear the PR backlog. I'll leave the branch open in case we want to revisit this

**Comment by ohld** [comment]
Any updates on implementing the heatmaps? I have the stream of geo points and the only way they can be visualized is Pin map...

**Pull request closed without merging**